pkg/provider/infisical/token.go

```diff
@@ -0,0 +1,170 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or impliec.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package infisical
+
+import (
+	"errors"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/external-secrets/external-secrets/pkg/provider/infisical/api"
+)
+
+type MachineTokenManager struct {
+	accessTokenTTL           time.Duration
+	accessTokenMaxTTL        time.Duration
+	accessTokenFetchedTime   time.Time
+	accessTokenRefreshedTime time.Time
+
+	mutex sync.Mutex
+
+	accessToken  string
+	clientSecret string
+	clientID     string
+
+	apiClient api.InfisicalApis
+}
+
+func NewMachineTokenManger(apiClient api.InfisicalApis, clientID, clientSecret string) *MachineTokenManager {
+	token := &MachineTokenManager{
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		apiClient:    apiClient,
+	}
+
+	return token
+}
+
+func (t *MachineTokenManager) StartLifeCycle() error {
+	if err := t.FetchNewAccessToken(); err != nil {
+		return err
+	}
+	go t.HandleTokenLifecycle()
+	return nil
+}
+
+func (t *MachineTokenManager) HandleTokenLifecycle() {
+	// this is skip doing things in first iteration.
+	// the refresh part is not needed to trigger at first because access token is just received as on top.
+	init := false
+	for {
+		accessTokenMaxTTLExpiresInTime := t.accessTokenFetchedTime.Add(t.accessTokenMaxTTL - (5 * time.Second))
+		accessTokenRefreshedTime := t.accessTokenRefreshedTime
+
+		if accessTokenRefreshedTime.IsZero() {
+			accessTokenRefreshedTime = t.accessTokenFetchedTime
+		}
+
+		if time.Now().After(accessTokenMaxTTLExpiresInTime) {
+			Logger.Info("Infisical Authentication: machine identity access token has reached max ttl, attempting to re authenticate...")
+			err := t.FetchNewAccessToken()
+			if err != nil {
+				Logger.Error(err, "Infisical Authentication: unable to authenticate universal auth. Will retry in 30 seconds")
+
+				// wait a bit before trying again.
+				time.Sleep((30 * time.Second))
+				continue
+			}
+		} else if init {
+			err := t.RefreshAccessToken()
+			if err != nil {
+				Logger.Error(err, "Infisical Authentication: unable to refresh universal auth token because %v. Will retry in 30 seconds")
+
+				// wait a bit before trying again.
+				time.Sleep((30 * time.Second))
```

Comment:
We for sure don't want an operator thread to stop for 30 seconds for this.

Reply:
@gusfcarvalho the whole refresh is inside a go routine and not the main thread. So is that an issue, our infisical operator follows the same logic too.

Reply:
🤔 I'm not sure. when we do concurrency, things are also in go routines and we've seen so many issues with parallel clients using each others' creds - due to caching - so we didn't set them anywhere almost on purpose.

If we follow with this, though, I would expect support for open issues on the controller level as well - as a blocking thread impacts not only this provider (but any other provider the user might be configuring).

Reply:
Haha. We are happy to support this amazing project. 

Let me go through as you recommended vault structure and also discuss it our team. Gonna ensure your concern is handled.

Reply:
Hey @gusfcarvalho 

I have changed to a more simple strategy to just use the token and then revoke it after use as you said. Its indeed a better and simple.